spiffe/id.go

```diff
@@ -0,0 +1,242 @@
+package spiffe
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type idType int
+
+const (
+	anyId idType = iota
+	trustDomainId
+	workloadId
+)
+
+// ValidateID validates the SPIFFE ID according to the SPIFFE
+// specification. The validation mode controls the type of validation.
+func ValidateID(spiffeID string, mode ValidationMode) error {
+	_, err := ParseID(spiffeID, mode)
+	return err
+}
+
+// ValidateIDURL validates the SPIFFE ID according to the SPIFFE
+// specification, namely:
+// - spiffe id is not empty
+// - spiffe id is a valid url
+// - scheme is 'spiffe'
+// - user info is not allowed
+// - host is not empty
+// - port is not allowed
+// - query values are not allowed
+// - fragment is not allowed
+// - path does not start with '/spire' since it is reserved for agent, server, etc.
+// In addition, the validation mode is used to control what kind of SPIFFE ID
+// is expected.
+// For more information:
+// [https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md]
+func ValidateIDURL(id *url.URL, mode ValidationMode) error {
+	options := mode.validationOptions()
+
+	validationError := func(format string, args ...interface{}) error {
+		var kind string
+		switch options.idType {
+		case trustDomainId:
+			kind = "trust domain "
+		case workloadId:
+			kind = "workload "
+		}
+		return fmt.Errorf("invalid %sSPIFFE ID %q: "+format,
```

Comment:
is there a missing space here?

Reply:
heh, no... the kind string has the trailing space. if the "kind" isn't set then it keeps the string sane.